src/core/api/api_access_control.pl

```diff
@@ -0,0 +1,486 @@
+              api_get_user_from_name/4,
+              api_get_user_from_id/4,
+              api_update_user_password/4
+          ]).
+
+:- use_module(core(util)).
+:- use_module(core(account)).
+:- use_module(core(query)).
+:- use_module(core(document)).
+:- use_module(core(transaction)).
+:- use_module(library(plunit)).
+:- use_module(library(crypto)).
+:- use_module(library(lists)).
+:- use_module(library(apply)).
+:- use_module(library(yall)).
+
+api_get_roles(SystemDB, Auth, Roles) :-
+    do_or_die(
+        is_super_user(Auth),
+        error(access_not_authorised(Auth,'Action/manage_capabilities','SystemDatabase'), _)),
+    get_roles(SystemDB, Roles).
+
+get_roles(SystemDB,Roles) :-
+    findall(Role,
+            ask(SystemDB,
+                (   t(Id, rdf:type, '@schema':'Role'),
+                    get_document(Id, Role)
+                )),
+            Roles).
+
+api_get_role_from_id(SystemDB, Auth, Id, Role) :-
+    do_or_die(
+        is_super_user(Auth),
+        error(access_not_authorised(Auth,'Action/manage_capabilities','SystemDatabase'), _)),
+    do_or_die(
+        get_role_from_id(SystemDB, Id, Role),
+        error(no_role_with_given_id(Id), _)
+    ).
+
+%% get_role_from_name(+,+,-) is det + error
+get_role_from_id(SystemDB, Id, Role) :-
+    ask(SystemDB,
+        (   t(Id, rdf:type, '@schema':'Role'),
+            get_document(Id, Role)
+        )).
+
+api_get_role_from_name(SystemDB, Auth, RoleName, Role) :-
+    do_or_die(
+        is_super_user(Auth),
+        error(access_not_authorised(Auth,'Action/manage_capabilities','SystemDatabase'), _)),
+    get_role_from_name(SystemDB, RoleName, Role).
+
+%% get_role_from_name(+,+,-) is det + error
+get_role_from_name(SystemDB, RoleName, Role) :-
+    findall(Role,
+            ask(SystemDB,
+                (   t(Id, name, RoleName^^xsd:string),
+                    t(Id, rdf:type, '@schema':'Role'),
+                    get_document(Id, Role)
+                )),
+            Roles),
+    (   Roles = [Role]
+    ->  true
+    ;   Roles = []
+    ->  throw(error(no_id_for_role_name(RoleName), _))
+    ;   throw(error(no_unique_id_for_role_name(RoleName),_))
+    ).
+
+api_add_role(_, Auth, Role, Id) :-
+    do_or_die(
+        is_super_user(Auth),
+        error(access_not_authorised(Auth,'Action/manage_capabilities','SystemDatabase'), _)),
+
+    % Make sure that we only insert role types.
+    put_dict('@type', Role, 'Role', Typed_Role),
+
+    create_context(system_descriptor{}, commit_info{author: "admin", message: "API: Add Role"},
+                   System_Context),
+    with_transaction(
+        System_Context,
+        insert_document(System_Context,Typed_Role,Id),
+        _
+    ).
+
+api_update_role(_, Auth, Role) :-
+    do_or_die(
+        is_super_user(Auth),
+        error(access_not_authorised(Auth,'Action/manage_capabilities','SystemDatabase'), _)),
+
+    % Make sure that we only insert role types.
+    put_dict('@type', Role, 'Role', Typed_Role),
```

Comment:
Same concern here regarding treating input data as documents to put into the system database. See above.